ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+
+An equivalent problem appears if you want to run a global query, such has the number of
+unique ip addresses who visited any site over some time period. Without HLLs this query
+involves shipping lists of ip addresses from the workers to the master for it to
+deduplicate. That's both a lot of network traffic and a lot of computation. By using HLLs
+you can greatly improve query speed.
+
+First you must install the hll extension; `the github repo
+<https://github.com/aggregateknowledge/postgresql-hll>`_ has instructions. Next, you have
+to enable it:
+
+.. code-block:: sql
+
+  -- this part must be run on all workers
+  CREATE EXTENSION hll;
+
+  -- this part runs on the master
+  ALTER TABLE http_requests_1min ADD COLUMN distinct_sessions (hll);
+
+When doing our rollups, we can now aggregate sessions into an hll column with queries
+like this:
+
+.. code-block:: sql
+
+  SELECT
+    zone_id, date_trunc('minute', ingest_time) as minute,
+    hll_add_agg(hll_hash_text(session_id)) AS distinct_sessions
+  WHERE minute = date_trunc('minute', now())
+  FROM http_request
+  GROUP BY zone_id, minute;
+
+Now dashboard queries are a little more complicated, you have to read out the cardinality
+during SELECT:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec,
+    hll_cardinality(distinct_sessions) AS distinct_session_count
+  FROM http_request_1min
+  WHERE zone_id = 1 AND minute = date_trunc('minute', now());
+
+HLLs aren't just faster, they let you do things you couldn't previously. Say we did our
+rollups, but instead of using HLLs we saved the exact unique counts. This works fine, but
+you can't answer queries such as "how many distinct sessions were there during this
+one-week period in the past we've thrown away the raw data for?". With HLLs, it's easy:
+
+.. code-block:: sql
+
+  -- careful, doesn't work!
+  SELECT
+    hll_cardinality(hll_union_agg(distinct_sessions))
+  FROM http_request_1day
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-08-2016';
+
+Well, it would be easy, except since Citus `can't yet
+<https://github.com/citusdata/citus/issues/120>`_ push down aggregates such as
+hll_union_agg. Instead you have to do a bit of trickery:
+
+.. code-block:: sql
+
+  -- this should be run on the workers and master
+  CREATE AGGREGATE sum (hll)
+  (
+    sfunc = hll_union_agg,
+    stype = internal,
+  );
+
+Now, when we call SUM over a collection of hlls, postgresql will return the hll for us.
+This lets us write the above query as:
+
+.. code-block:: sql
+
+  -- working version of the above query
+  SELECT
+    hll_cardinality(SUM(distinct_sessions))
+  FROM http_request_1day
+  WHERE ingest_time BETWEEN timestamp '06-01-2016' AND '06-08-2016';
+
+More information on HLLs can be found in `their github repo
+<https://github.com/aggregateknowledge/postgresql-hll>`_.
+
+HLLS are an example from the postgresql community, but there are a couple extensions we've
```

Comment:
We are part of the postgresql community ;). 

Maybe something like: 
The HLL extension only provides approximate distinct counts, but there are other extensions which do similar approximations for different kinds of queries, this includes ...